メッセージ一覧機能の追加

sya-ri opened the pull request:
- ページの追加
  - `/` タイムライン

    ![image](https://user-images.githubusercontent.com/34268371/133027727-82012716-f24d-4196-9a02-00bacbdfa3ed.png)
  - `/message/{id}` メッセージの表示

    ![image](https://user-images.githubusercontent.com/34268371/133027764-ee396a28-70f5-4eba-8d9e-59081685282e.png)

- "いいね" の一時的な追加
  - 🖤, ❤️ をクリックすると [LikeButton#onClick](https://github.com/sya-ri/t5r/blob/9f4c0f8badda12611be034ecba6f78d7522b2ad1/app/Http/Livewire/LikeButton.php#L21) が呼び出されます
  - 🖤 をクリックすると ❤️ になります(戻りません)
  - 🖤, ❤️ をクリックするとメッセージの以下の要素が再レンダリングされます
    - いいね数
    - いいねの表示

sya-ri committed:
feat: モデルに Message を追加
- migration にテーブルの定義を追加
- MessageSeeder を追加
- MessageFactory を追加
- UserSeeder を追加

参考: https://qiita.com/sutara79/items/9e57e7af03b65e0d6047

sya-ri committed:
feat: TimelineController を追加
dashboard という名前を timeline に変更

sya-ri committed:
feat: タイムラインにメッセージ一覧を表示

sya-ri committed:
change: メッセージ一覧のUIを改善
- スタイルを追加
- メッセージをクリックすると /message/{id} にアクセスする
- いいねボタンのクリック可能にした

sya-ri committed:
feat: /message/{id} のページを作成

sya-ri committed:
chore: livewire を導入

sya-ri committed:
feat: いいねボタンを押せるようにした
refactor: いいねボタンのコンポーネント化

sya-ri committed:
change: モバイルでスタイルが崩壊する問題を修正

sya-ri committed:
fix: 存在しないユーザーのメッセージを作成する問題を修正

sya-ri committed:
fix: 存在しないメッセージにアクセスした時に 500 エラーが出る問題を修正

sya-ri committed:
test: メッセージ一覧機能のテストを追加

d2verb commented on app/Http/Controllers/MessageViewController.php:
```diff
@@ -0,0 +1,20 @@
+namespace App\Http\Controllers;
+
+use App\Models\Message;
+use Illuminate\Http\Request;
```
(nits)
Request はこのファイル内で使われていないです。消しても問題ないかと。

d2verb commented on app/Http/Controllers/MessageViewController.php:
```diff
@@ -0,0 +1,20 @@
+        }
+
+        abort(404, 'Message Not Found');
+    }
```
(suggest)
Laravel のルートモデル結合を使うとスッキリします。（web.php も変更する必要があります）
https://readouble.com/laravel/8.x/ja/routing.html#route-model-binding

```suggestion
    public function show(Message $message)
    {
        return view('message-view', compact('message'));
    }
```

d2verb commented on database/factories/MessageFactory.php:
```diff
@@ -0,0 +1,32 @@
+    public function definition()
+    {
+        return [
+            'user_id' => User::all()->random(),
```
(suggest)
これだと何かしらのユーザーが存在する前提になりませんか？（MessageFactory を実行する前に必ず User が DB に存在することを MessageFactory の利用者が意識しないといけない）

以下のようにすれば、MessageFactory 実行時に勝手に User も作成されます。
```suggestion
            'user_id' => User::factory(),
```

d2verb commented on database/migrations/2021_09_09_073703_create_messages_table.php:
```diff
@@ -0,0 +1,37 @@
+    {
+        Schema::create('messages', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedBigInteger('user_id');
+            $table->foreign('user_id')
+                ->references('id')->on('users')
+                ->cascadeOnDelete();
```
(suggest)
Laravel 8 から以下のように短く書けます！
see: https://readouble.com/laravel/8.x/ja/migrations.html#foreign-key-constraints
```suggestion
            $table->foreignId('user_id')->constrained()->onDelete('cascade');
```

d2verb commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,38 @@
+    {
+        $user = User::factory()->create();
+
+        $response = $this->actingAs($user)->get('/');
```
(imo)
パスを直接書くのではなく、ルートの名前を指定したほうが変更に強いかなと思います。

```suggestion
        $response = $this->actingAs($user)->get(route('timeline'));
```

d2verb commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,38 @@
+
+class MessageTest extends TestCase
+{
+    public function test_timeline_screen_can_be_rendered()
```
(imo)
メッセージの一覧が表示されていることも検証したほうがよいかと思います。このままでは空のページでもテストが通ってしまいますからね。
see: https://readouble.com/laravel/8.x/ja/http-tests.html#testing-views

d2verb commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,38 @@
+        $response->assertStatus(200);
+    }
+
+    public function test_message_screen_can_be_rendered()
```
(imo)
>メッセージの一覧が表示されていることも検証したほうがよいかと思います。このままでは空のページでもテストが通ってしまいますからね。
see: https://readouble.com/laravel/8.x/ja/http-tests.html#testing-views

ここも↑と同様です。

d2verb commented on database/factories/MessageFactory.php:
```diff
@@ -0,0 +1,32 @@
+    {
+        return [
+            'user_id' => User::all()->random(),
+            'content' => Str::random(60),
```
(suggest)
こっちのほうがよりそれっぽい文章が生成されるかと思います。

```suggestion
            'content' => $this->faker->realText(60),
```

Yumemi-TOnodera commented on app/Http/Controllers/MessageViewController.php:
```diff
@@ -0,0 +1,20 @@
+            return view('message-view', compact('message'));
+        }
+
+        abort(404, 'Message Not Found');
```
(suggest)
- 404 がマジックナンバーになっているので、こちらの記事が参考になるかも。 （他の箇所にも適用を）https://codelikes.com/laravel-const-http-status-code/

sya-ri replied:
`assertStatus` と `Symfony\Component\HttpFoundation\Response` を使うのではなく、`assertOk` 等のメソッドを使う方が綺麗だと思ったので修正しました。

https://github.com/sya-ri/t5r/pull/4/commits/203c01c5620ec8aa8d0ad17d2f75224471bdba5a

Yumemi-TOnodera commented on app/Http/Controllers/TimelineController.php:
```diff
@@ -0,0 +1,20 @@
+{
+    public function show()
+    {
+        $messages = Message::all([
```
(suggest)
- ここのように明らかに将来的に問題になりそうな箇所は、TODOとしておきましょう。

```suggestion
        // TODO: offsetやlimitの指定が将来的に必要になりそう（ページングやレイジーロードがの検討）
        $messages = Message::all([
```

sya-ri committed:
refactor: 不要なインポートを削除
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707104066

sya-ri committed:
refactor: ルートモデル結合を使って処理の簡略化をした
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707115702

Yumemi-TOnodera commented on app/Http/Livewire/LikeButton.php:
```diff
@@ -0,0 +1,26 @@
+    public $message;
+    public $likes;
+    public $icon;
+    private $isLike; // いいねしているか TODO いいね機能実装時に削除
```
(should)
初期化タイミングがないので。

```suggestion
    private $isLike = false; // いいねしているか TODO いいね機能実装時に削除
```

sya-ri replied:
確かに真偽値として値を確定させておくことは重要ですね。そうするべきでした。#5 でその行は削除されるので、この PR では変更を加えないでマージしたいです。

https://github.com/sya-ri/t5r/blob/2f0b095e2c8fef957c6155814a976029f1d53b7e/app/Http/Livewire/LikeButton.php#L11-L32

Yumemi-TOnodera replied:
了解！

Yumemi-TOnodera commented on app/Models/Message.php:
```diff
@@ -0,0 +1,27 @@
+
+    public function likes()
+    {
+        return rand(0, 100);
```
(ask)
- ここは TODO なコード？

sya-ri replied:
その通りです。 TODO をつけておくべきでした。いいね機能がなければいいね数が取得できないので、代わりに rand を使っています。メソッド名は変わっていますが、#5 で以下のように実装されているので、この PR では TODO コメントをつけずにマージしたいです。
https://github.com/sya-ri/t5r/blob/2f0b095e2c8fef957c6155814a976029f1d53b7e/app/Models/Message.php#L23-L26

Yumemi-TOnodera replied:
了解！

sya-ri committed:
change: UserがひとりもいなくてもMessageFactoryを使えるようにした
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707119916

Yumemi-TOnodera reviewed (commented):
レビューしました

sya-ri committed:
refactor: テーブル定義の簡略化をした
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707143259

sya-ri committed:
test: ルートパスを動的に取得するようにした
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707165135

sya-ri committed:
change: 自動生成されるメッセージ内容を実際の文章に近くした
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707171483

sya-ri committed:
docs: 今後変更を加えるべき箇所に TODO を追加

指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707182308

Co-authored-by: Takahiro Onodera @ Yumemi Inc. <45787230+Yumemi-TOnodera@users.noreply.github.com>

sya-ri committed:
refactor: マジックナンバーになっているステータスコードを修正
指摘: https://github.com/sya-ri/t5r/pull/4#discussion_r707179868

d2verb commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,39 @@
+        $user = User::factory()->create();
+        $message = Message::factory()->create();
+
+        $response = $this->actingAs($user)->get("/message/$message->id");
```
(suggest)
```suggestion
        $response = $this->actingAs($user)->get(route('message.view', [$message->id]));
```

d2verb commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,39 @@
+    {
+        $user = User::factory()->create();
+
+        $response = $this->actingAs($user)->get("/message/-1"); // -1 is an unexpected id (id is unsigned)
```
(suggest)
```suggestion
        $response = $this->actingAs($user)->get(route('message.view', [-1])); // -1 is an unexpected id (id is unsigned)
```

d2verb reviewed (commented):
レビュー漏れがあったので追記しました。パスを直書きしている箇所です。

Yumemi-TOnodera reviewed (dismissed):
別PRで修正予定との部分把握しました。

sya-ri committed:
refactor: assertStatus を直接使わないように変更

sya-ri committed:
test: メッセージの内容が表示されているかどうかを確認するようにした
指摘[1]: https://github.com/sya-ri/t5r/pull/4#discussion_r707167330
指摘[2]: https://github.com/sya-ri/t5r/pull/4#discussion_r707167809

sya-ri committed:
test: ルートパスを動的に取得するようにした (more fix)
指摘[1]: https://github.com/sya-ri/t5r/pull/4#discussion_r707213402
指摘[2]: https://github.com/sya-ri/t5r/pull/4#discussion_r707214125

Yumemi-TOnodera reviewed (approved):
一旦LGTMにしておきますねー

sya-ri commented on tests/Feature/MessageTest.php:
```diff
@@ -0,0 +1,45 @@
+        $response->assertSee($message->content);
+    }
+
+    public function test_message_is_not_found()
+    {
+        $user = User::factory()->create();
+
+        $response = $this->actingAs($user)->get("message.view", [-1]); // -1 is an unexpected id (id is unsigned)
+
+        $response->assertNotFound();
+    }
```
`route` をつけ忘れており、header に `-1` が指定され、`/message` にアクセスしています。しかし、`/message` というパスは存在しないのでテスト通過しています。`assertNotFound` は無駄でしかないことに気がついたので、 #6 でテストを追加する際に削除します。今回のPRでは無視してください。

~~変更したコミット: https://github.com/sya-ri/t5r/pull/6/commits/c3179f73a64ad9d1201643df31872a33997c531c~~

@d2verb [指摘いただいた内容](https://github.com/sya-ri/t5r/pull/4#discussion_r707214125)に関係します。よろしくお願いします

d2verb replied:
うーむ、それでもこのテストはあったほうがいいかなと思います。
というのも、このテストがあれば `route("message.view", [-1])` が 200 を返すという不具合を検知できるからです。
今後デグレが発生したときのために残しておいたほうがいいと思います。

sya-ri replied:
なるほど。どうミスをすればそういうバグが起こるのか検討もつきませんがテストで失敗することも確認しておいた方が安心できますね。わかりました。修正してこちらのPRにあげておきます。

sya-ri replied:
こちら、修正しました
https://github.com/sya-ri/t5r/pull/4/commits/fdcc84b17cd236eb5d6ad30621537b66090e81bc

sya-ri committed:
test: route の指定を間違っている問題を修正
ref: https://github.com/sya-ri/t5r/pull/4#discussion_r707947412
